Rewrite rewrite

atomcult opened the pull request:
An assortment of fixes and an initial rewrite of auto-mount

Commit:
auto-install: fix check_drive() loop

Commit:
auto-install: split out shared variables and functions to scripts/bin/common

Commit:
snapcraft.yaml: change udisksd installation prefix to /usr

Commit:
auto-mount: initial rewrite of auto-mount

Commit:
use fifo for signaling device automount

Commit:
appease shellcheck

Commit:
auto-mount: make mount_devices() function guard more robust

dilyn-corner commented:
This looks quite good and I appreciate the effort!

We'll have to test but I don't see anything obviously wrong. Thanks for also jumping ahead and fixing the things I was going to do after the initial rewrite!

Commit:
auto-mount: replace code using udisks

atomcult commented:
No problem!

Tacked on a bit more that replaces all the code that uses udisks with code that uses inotifywatch, but I haven't yet removed any of the dependencies from the snapcraft.yaml.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,52 @@
+        while read -r _parent _action _uuid; do
+            case "$_action" in
+                MOVED_TO) try_mount "$_parent" "$_uuid" ;;
+                DELETE)   rm_mount  "$_uuid" ;;
```
preemptive `rm_mount not found`

dilyn-corner replied:
Note from conversation: unmounting is hard to do in a clean fashion (no good way of indicating a "finished" state). What we can instead do it is unmount the partition and trigger a system reboot, which is a fine enough indication that the process has finished.

We can leave a comment indicating other options one can take, but we shouldn't implement them.

atomcult replied:
We should think about what to do in cases where no snaps are imported. Unfortunately, the best option is probably to simply avoid rebooting in that case, otherwise we get reboots even if we're just trying to import user assertions.

atomcult replied:
Also, this could potentially be a DoS attack, if someone just leaves in a USB drive 😂

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,52 @@
+            case "$_action" in
+                MOVED_TO) try_mount "$_parent" "$_uuid" ;;
+                DELETE)   rm_mount  "$_uuid" ;;
+            esac
```
We should have a third case, even if it's just `*) ;;`

atomcult replied:
That's a good point. In theory, that should exhaust all the options, since we're only listening for `moved_to` and `delete` events, but who knows what other garbage could get thrown our way, especially if there's an error

dilyn-corner replied:
If we get an error, we just `panic() { panic | panic }`

atomcult replied:
The "If I can't run, nobody can" method of error handling. I like it

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,52 @@
+    set -eu
+
+    # TODO: Should there be a $_log_dir env var in /bin/common?
+    _date=$(date -Iseconds); readonly _date
```
Could just `readonly _date=$(date -ISeconds)`, correct?

atomcult replied:
I just do what shellcheck tells me to do
https://www.shellcheck.net/wiki/SC2155

dilyn-corner replied:
I like intentionally obfuscating return values though :(

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,52 @@
+        return
+    fi
+
+    echo "${_target}" >> "${_watch_file}"
```
So `>>` instead of `>` would mean that we support installing snaps from more than just one partition. It's not wrong, but I'm not certain that's what we want?

atomcult replied:
I'm actually not sure there's a difference with a FIFO. When I originally wrote this, I wasn't thinking about the fact that writing to a FIFO blocks until read from, so I wanted to make sure that each line was appended. I think it's completely reasonable to switch this over to `>`.

Anyhow, try_mount() is always called with a single device as an argument, and every device that's newly populated in `/dev/disk/by-uuid` will be processed, so by default we are actually checking every partition.

dilyn-corner replied:
_it's a fifo_ 
Then this is fine and it's overwhelming pedantry.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,52 @@
+main() {
+    set -eu
+
+    # TODO: Should there be a $_log_dir env var in /bin/common?
```
I think if we're going to have a `common` and we're putting at least one variable in it, there isn't a strong reason against putting every variable in it (at least the _common_ ones).

dilyn-corner commented on scripts/bin/common:
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
```
We don't need a shebang (or `+x`) if it's just being sourced.

atomcult replied:
I used a shebang rather than have vim magic at the top to set the filetype, since it's more universal.

dilyn-corner replied:
What do I do when a person exhibits undefined behavior

Commit:
tests: add basic unit test framework

Commit:
udisksd: remove patch and script

Commit:
auto-install: simplify find_assertions()

Commit:
auto-install: get json response via subshell instead of output file

Commit:
tests: make test script more generic, add find_assertions test

Commit:
tests: improve find_assersions test

Commit:
test: fix assert_eq() and assert_neq()

Commit:
auto-install: rewrite functions so that calls aren't chained

Commit:
auto-mount: use mklog() instead of init_log()

Commit:
auto-mount: various fixes and cleanup

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -73,92 +40,112 @@ rest_call() {
+    # keep trying to set the channel to stable because it may take some
+    # if a snap change is in proces give up after 14 loops, where
+    # each loop sleeps twice as long as the previou where 14 loops
+    # takes 19.1 minute
```
I think it's 17.03 minutes

I don't think there's actually a way to fix this; we *have* to communicate with the store and get a response for this to succeed, and there's no way of making that faster or cleaner afaik

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -73,92 +40,112 @@ rest_call() {
+        [ -d "$_mount_point" ] || continue
+
+        # Skip mounts that don't contain any asserts
+        ls "$_mount_point"/* >/dev/null 2>&1 || continue
```
Should be `ls "$_mount_point"/*.assert`?
Maybe this should be a `find` instead of an `ls`; supporting arbitrarily nested asserts and snaps is trivial if we use `basename`.

atomcult replied:
Yes lol. You would not believe how many times I have made that mistake.
And yeah, I think using `find` would be a good improvement

dilyn-corner commented:
I can't assign anyone here, but @mboudr35 your thoughts are appreciated :)

Commit:
auto-install: fix comment for track_stable()

Commit:
auto-mount: temporarily fix *.assert enumeration

Commit:
auto-mount: fix NOEXEC conditional

Commit:
auto-install: fix curl invocations

Commit:
auto-mount: ensure block dev's mount dir is created

dilyn-corner commented on udisksd.patch:
```diff
@@ -1,83 +0,0 @@
-diff --git a/src/udisksdaemon.c b/src/udisksdaemon.c
```
For posterity: we don't require this patch anymore because...?

atomcult replied:
Because we're not using udisksd anymore. Snapcraft.yaml still doesn't reflect that, but I figured I'd go ahead and clean up the patch and the udisks script

dilyn-corner replied:
Just in time for us to officially have a udisks2 snap *again*!

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -1,70 +1,37 @@
-    _socket=/run/snapd.socket
+    _file="$2"
+
+    _snap_name="$(basename "${2%%_*}")"
```
Do these quotes do what you think they do?
I think we can skip the inner "" on the substitution.

atomcult replied:
I dunno, what do you think I think they do?

dilyn-corner replied:
I don't think you think they do anything :v 

Quotes are vile... I think in the case of `$(...)` nested quotes are fine though...? Not certain. At the very least they'd be safe to remove, but I guess I don't care.

atomcult replied:
Yeah, quotes are treated specially within command substitutions. And I think they *are* necessary, since `${2%%_*}` could contain spaces.

dilyn-corner replied:
It better not!

ETA: ... Ugh. Yes, yes.

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -73,92 +40,110 @@ rest_call() {
+        [ -d "$_mount_point" ] || continue
+
+        # Skip mounts that don't contain any asserts
+        ls "$_mount_point"/*.assert >/dev/null 2>&1 || continue
```
Is this actually just busy-waiting for a mount to happen? If the mount point exists but there aren't any found assertions at that mount point, I think we just log and exit (there's nothing to do)?

atomcult replied:
Ah, yeah, we should be copying the log over and doing all the unmount stuff even if there are no asserts. Good catch.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,50 @@
+}
+
+try_mount() {
+    _parent="${1%/}"
```
Odd nit: technically we don't have to do a substitution to strip the lagging `/` because (relevant enough) versions of libc collapse particular repetitions of `/` to the expected result. I believe in all cases, `//` is collapsed to `/`. 

I think you only did this to make the code nicer to look at, and I *doubt* there's any actual problem with doing this, but I felt like showing off.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,50 @@
+}
+
+rm_mount() {
+    # TODO
```
I think `rm_mount()` is just trivially `umount $_target && shutdown {-h|-H $message}`, but of course the Store team recently made the required interface for powering off a system super-privileged. So I think we may just scrap the method of notifying...

atomcult replied:
Agreed, although we should `sync`, `umount`, and `eject` in `auto-install`, since it's the script that actually knows when all the snaps are done being installed. Once `auto-install` unmounts the drive, inotifywatch will emit a `DELETE` message for that device, and we could optionally remove the directory that it was mounted in. By default, that's in a temporary directory, so it normally shouldn't matter, but `MOUNT_DIR` *is* overrideable.

dilyn-corner replied:
Sure, if you want to be *safe about it* :P 

There's no down-side to cleaning up after ourselves even if it's just in `/tmp`; it's probably the correct move because (as you said) `MOUNT_DIR` could be set somewhere else.

dilyn-corner commented on scripts/bin/common:
```diff
@@ -0,0 +1,17 @@
+
+    [ -d "$_log_dir" ] || mkdir -p "$_log_dir"
+
+    echo "${_log_dir}/$(date -Iseconds).log"
```
`echo` -> `touch`?

atomcult replied:
As it's currently used, `mklog()` is called just to generate an appropriate name, and is implicitly created when `log()` is called. That beings said, I do think it's probably worth initializing empty logs, though. An empty log is definitely better than no log.

atomcult replied:
As it's currently used, mklog() is called just to generate an appropriate name, and is implicitly created when log() is called. That beings said, I do think it's probably worth initializing empty logs, though. An empty log is definitely better than no log.

dilyn-corner replied:
I think it would be good to create the empty log because what happens if we can't actually write to the filesystem (imagine an SD card with write-lock on).

dilyn-corner commented on test:
```diff
@@ -0,0 +1,76 @@
+}
+
+dirname() {
+    _dir="${1%/}"
```
Do we want to strip all trailing `/` or just the last one?

Compare:

```
_dir=foo///
echo ${_dir%/}
echo ${_dir%%"${_dir##*[!/]}"}
```

Not certain if it matters but I also don't know why implement `dirname` in `sh` if we already have `/usr/bin/dirname`.

dilyn-corner commented on test:
```diff
@@ -0,0 +1,76 @@
+}
+
+main() {
+
```
How dare you put this blank line here

atomcult replied:
Well, I have to inflate my line counts somehow :P

dilyn-corner reviewed (commented):
Left some extra comments. I think we're very close now!

Commit:
test: remove unnecessary dirname() function

Commit:
auto-mount: remove mount dir after device is ejected

Commit:
bin/common: create log file within mklog()

Commit:
auto-install: eject devices once we're done with them

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -101,6 +101,15 @@ track_stable() {
+
+    sync
+    umount "$_mount_point"
+    eject "/dev/disk/by-uuid/$(basename "$_mount_point")"
```
Remake to `eject_device $WATCH_FILE` with `_disk="$1` and `eject $_disk`, or something like?

atomcult replied:
Is your issue with implicitly passed arguments (`$_mount_point`) or reverse engineering the block device path?

I don't really like implicitly passing args, but my instinct was to have `_mount_point="$1"`, which felt even worse. Other options are to use a different arg for (maybe just `_mount`) or we could also make the function run in a subshell:

```sh
eject_device() (
  _mount_point="$1"
  
  # ...
)
```

dilyn-corner replied:
My thinking was that because we already have a FIFO which holds the path to the partition we can just reuse that

ETA
I'm just now realizing that I misread this block.

`eject` seems to target the entire block device, not just some partition. So the target should be the resolution of the link at `/dev/disk/by-uuid/foo` without the partition number.

dilyn-corner replied:
> [If a device partition is specified, the whole-disk device is used.](https://www.man7.org/linux/man-pages/man1/eject.1.html)

Nevermind!

Though I think because we have a mounting script we could probably do with moving the eject logic to their and having this script call that script to do the work... Added benefit of merely getting to reuse `$MOUNT_POINT`...

atomcult replied:
Hmm.. that's a good point. Depending on how complicated this ends up being, we could potentially live dangerously and forego ejecting entirely. As it stands, users have to just guess when the drive has been ejected anyway, unless they implement custom notification functionality.

dilyn-corner replied:
I think ejecting is the minimum safety level considering we do write to the disk. I'm thinking that we should have an idea surrounding what cleanup looks like which can be extended in the future (i.e. notifications).

dilyn-corner replied:
That being said, I think I'm satisfied by this PR at this point.

Commit:
auto-install: wait for fifo

dilyn-corner committed:
Appease shellcheck

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
auto-install: use find to allow arbitrary nesting of assertions

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

Commit:
scripts: make instances of `return` have explicit values

dilyn-corner committed:
snapcraft.yaml: bump to core22 base

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapcraft.yaml: drop udisks2

Mount functionality no longer relies on udisks2 so we can safely remove
it from the snap. We replace the udisks2 interface with the
mount-control interface, which can be extended or restricted as much as
needed.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
auto-mount: remove target dir on mount failure

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
auto-mount: uniformally declare variables readonly

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

Commit:
auto-install: allow unmounting

Commit:
snapcraft.yaml: remove e2fsprogs, add eject

Commit:
auto-mount: add mount options and default fstype

Commit:
scripts: add cleanup functions, move mgmt of WATCH_FILE to auto-install

Commit:
scripts: minor fixes

Commit:
scripts: remove `scripts/bin/common`

Commit:
scripts: improve cleanup functions

Commit:
auto-mount: replace errant _mount_dir var

Commit:
auto-mount: allow auto-install to remove mount dirs

dilyn-corner committed:
auto-mount: bail if target dir is a mountpoint

Somehow, the symlinks in /dev/disk/by-uuid have the potential of being
created multiple times in rapid succession. Because this symlink has the
same name both times, we will end up attempting to mount the same
partition to the same location twice. The first time will succeed just
fine and auto-install will proceed to install all of the snaps. After a
short amount of time, however, try_mount() will attempt to resolve the
pending mount of the already mounted partition to the already mounted-on
target directory. Because the target directory is already a mountpoint,
the mount operation will fail.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -1,70 +1,54 @@
+mklog() {
+    # FIXME: if mklog() is called within a subshell,
+    #        then LOG_DIR isn't getting globally updated
+    _log_dir="${1:-"${LOG_DIR:="$(mktemp -d)"}"}"
```
We never call `mklog` with any args; is this a forward-looking thing and if so, what is the intention?

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -1,70 +1,54 @@
-
-    log "Mountpoint is ${_mount_point}."
-    find_assertions "${_mount_point}"
+    _log_path="${_log_dir}/$(date -Iseconds).log"
```
`_log_path` feels like a misnomer as `_log_dir` should already be the path. Should we use `_log_file` instead, with it being just the name and have `_log_path` be `${_log_dir}/${_log_file}`?

dilyn-corner commented on scripts/bin/auto-install:
```diff
@@ -73,92 +57,142 @@ rest_call() {
+    cp -f "$_log_file" "$_mount_point" || true
+
+    # Remove the temp directory
+    rm -r "$(dirname "$_log_file")"
```
If we take my above comment about `$_log_file` we can change this to `rm -rf $_log_dir` - this is a reminder note :)

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+}
+
+watch_devices() {
+    inotifywait --quiet --monitor --event "moved_to" --event "delete" /dev/disk/by-uuid | \
+        while read -r _parent _action _uuid; do
+            case "$_action" in
+                MOVED_TO) try_mount "$_parent" "$_uuid" ;;
+                *) : ;;
```
I don't think we need to watch the 'delete' event if we only have the `MOVED_TO` case.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+
+    mkdir -p "$_target"
+    grep -q "$_target" /proc/mounts || {
+        if ! mount -o rw,sync -t "${FSTYPE:-vfat}" "$_source" "$_target" >/dev/null 2>&1; then
```
Should we be `/dev/null`ing the output of this command? It seems relevant for e.g. debugging.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+    grep -q "$_target" /proc/mounts || {
+        if ! mount -o rw,sync -t "${FSTYPE:-vfat}" "$_source" "$_target" >/dev/null 2>&1; then
+            echo "WARN: Could not mount ${_source} to ${_target}"
+            rm -rf "$_target"
```
I think this should be `rmdir -p "$_target"` to avoid my fun issue of wiping my whole USB stick.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+
+    sync
+
+    _uuids="$(ls -1 "$MOUNT_DIR" | tr '\n' :)"
```
I dislike parsing `ls` but `find` here is... unpleasantly unsupported by POSIX...
I think this stays but we should remember to resolve this; mark with TODO

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+    sync
+
+    _uuids="$(ls -1 "$MOUNT_DIR" | tr '\n' :)"
+    if [ -n "$_uuids" ]; then
```
Pedantic: `[ -z "$_uuids" ] || { ... }`, but I don't mind.

dilyn-corner commented on scripts/bin/auto-mount:
```diff
@@ -1,17 +1,80 @@
+        IFS=:
+        for _uuid in $_uuids; do
+            umount "${MOUNT_DIR}/${_uuid}" || continue
+            rmdir "${MOUNT_DIR}/${_uuid}"
```
Should we use `rmdir -p "${MOUNT_DIR}/${_uuid}"` to ensure the temp dir is cleaned for this case?

dilyn-corner replied:
I guess it doesn't matter because we `rmdir "$MOUNT_DIR"` on line 63 :v

dilyn-corner reviewed (commented):
I think we're done here. I've left some comments but none of them block merging this.

I'm going to go through and leave some TODOs and potentially add some helpful comments to document things. After I've gotten some +1s we can merge :)

dilyn-corner committed:
scripts: add some documentation

All functions include a short explanation of their purpose and potential
limitations.

TODOs have been added to some pain points in the current code with
potential examples for implementation or references to documentation.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
snapcraft.yaml: document optional environment

Some variables used by auto-{install,mount} are modifiable in the
environment. These variables are now documented with their default
values specified as examples.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

Commit:
auto-mount: return early if mount already exists

Commit:
scripts: minor fixes and style changes

dilyn-corner committed:
auto-mount: enable environment settable options

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>
